src/compiler/checker.ts

```diff
@@ -3018,12 +3129,106 @@ namespace ts {
+                const chain = lookupSymbolChain(symbol, context, meaning);
+
+                if (expectsIdentifier && chain.length !== 1
+                    && !context.encounteredError
+                    && !(context.flags & NodeBuilderFlags.AllowQualifedNameInPlaceOfIdentifier)) {
+                    context.encounteredError = true;
+                }
+                return createEntityNameFromSymbolChain(chain, chain.length - 1);
+
+                function createEntityNameFromSymbolChain(chain: Symbol[], index: number): EntityName {
+                    const typeParameterNodes = lookupTypeParameterNodes(chain, index, context);
+                    const symbol = chain[index];
+                    const symbolName = getNameOfSymbol(symbol, context);
+                    const identifier = setEmitFlags(createIdentifier(symbolName, typeParameterNodes), EmitFlags.NoAsciiEscaping);
+                    identifier.symbol = symbol;
+
+                    return index > 0 ? createQualifiedName(createEntityNameFromSymbolChain(chain, index - 1), identifier) : identifier;
+                }
+            }
+
+            function symbolToExpression(symbol: Symbol, context: NodeBuilderContext, meaning: SymbolFlags) {
+                const chain = lookupSymbolChain(symbol, context, meaning);
+
+                return createExpressionFromSymbolChain(chain, chain.length - 1);
+
+                function createExpressionFromSymbolChain(chain: Symbol[], index: number): Expression {
+                    const typeParameterNodes = lookupTypeParameterNodes(chain, index, context);
+                    const symbol = chain[index];
+
+                    const symbolName = getNameOfSymbol(symbol);
+                    const firstChar = symbolName.charCodeAt(0);
+                    const needsElementAccess = !isIdentifierStart(firstChar, languageVersion);
+                    if (index !== 0 && needsElementAccess) {
+                        let expression: Expression;
+                        if (isSingleOrDoubleQuote(firstChar)) {
+                            expression = createLiteral(symbolName.substring(1, symbolName.length - 1).replace(/\\./g, s => s.substring(1)));
+                            (expression as StringLiteral).singleQuote = firstChar === CharacterCodes.singleQuote;
+                        }
+                        if (!expression) {
+                            // TODO: this matches legacy behavior, but results in syntactically invalid identifiers for numeric symbols and computed property name symbols
+                            expression = setEmitFlags(createIdentifier(symbolName, typeParameterNodes), EmitFlags.NoAsciiEscaping);
+                            expression.symbol = symbol;
+                        }
+                        return createElementAccess(createExpressionFromSymbolChain(chain, index - 1), expression);
+                    }
+                    else {

```

Comment:
I would flip this to put the common case first, something like `if (isLegalPropertyAccess || index === 0) ...`